### ISSUE TITLE
Exception thrown in the middleware stack crashes CAS sessions

### ISSUE THREAD

**Issue by borys-p** [subfission/cas#36]
Hi,

I'm using Explicit Route Bindings in my app and wanted to test custom 404 pages. And I kept getting those errors:
`phpCAS error: phpCAS::client(): ErrorException: session_start(): Cannot send session cache limiter - headers already sent (output started at (...)\vendor\symfony\http-foundation\Response.php:406) in (...)\vendor\subfission\cas\src\Subfission\Cas\CasManager.php on line 61`

It took me a while to pinpoint this issue. It happens because `Illuminate\Routing\Middleware\SubstituteBindings` throws `ModelNotFoundException` in `Illuminate\Routing\RouteBinding` and it's rendered as an HTML 404 page. And because this middleware is executed before `CASAuth`, `session_start()` that uses native PHP sessions crashes. You can easily test it by creating an middleware that raises an exception and putting it before `CASAuth`.

I can't move `CASAuth` middleware to the `web` stack, becase I need to have some routes accessible without SSO. For now I have extracted the `bindings` middleware into my `web` routes, but this needs a better solution, because an exception can happen in other places.

**Comment by borys-p** [comment]
This could help: https://github.com/Jasig/phpCAS/issues/42 but it's probably years away.

**Comment by subfission** [comment]
Have you tried wrapping the call to CAS inside the CAS middleware with a try/catch block that handles this exception?

There are at least a few ways to detect this type of issue and then either disregard the error or ensure the application will not trigger it.  I would recommend cloning the included sample middleware and then heavily modifying it for your project.

**Comment by borys-p** [comment]
The original exception happens before phpCAS middleware is executed - that's the whole problem. It happens in the routing middleware. And because I have custom 404 pages the exception is returned to the browser as an HTML page and **then** the request is passed to the phpCAS middleware. At this point session cannot be started, as headers were already sent and we have a fatal error. There is no exception to catch here.

**Comment by borys-p** [comment]
Hi, has it been fixed?

**Comment by subfission** [comment]
Hi boris-p,

I attempted your condition with one of my apps that also uses custom 404 pages and has CAS auth middleware active.  I do not have the same issue as you are having and I noticed that I catch fatal errors in the middleware.  

If you would like to send me your code for your middleware and routes, I can take a look, but it appears to be unique to your project and not this plugin.  I'm also curious as to whether your 404 pages are the culprit.
